### PR TITLE
Write issue taskq shouldn't be dynamic

### PR DESCRIPTION
**Pull request by dweeezil**
This is as much an upstream compatibility as it's a bit of a performance
gain.

The illumos taskq implemention doesn't allow a TASKQ_THREADS_CPU_PCT type
to be dynamic and in fact enforces as much with an ASSERT.

As to performance, if this taskq is dynamic, it can cause excessive
contention on tq_lock as the threads are created and destroyed because it
can see bursts of many thousands of tasks in a short time, particularly
in heavy high-concurrency zvol write workloads.

### CODE REVIEW

**Comment by dweeezil** [comment]
I'd like to follow this up with a patch to fan out the write issue taskq on systems without 16K stacks.  The zio dispatching in ```__zio_execute()``` can dramatically increase the contention on ```tq_lock```, particularly with the same type of zvol workloads described in the commit message for this PR.

**Comment by behlendorf** [comment]
@dweeezil do you have any performance or profiling data you could share on how bad this is?

**Comment by dweeezil** [comment]
@behlendorf I'm trying to get some solid numbers with lockdep and will post them when available.

tl;dr back-story:  My initial observations were based on watching ```perf top -g``` and seeing 50-70% CPU utilization from raw_spin_lock_irq as called from taskq_thread() and one other place.  Also from monitoring ```/proc/spl/taskq``` an seeing in some cases thousands and even hundreds of thousands of pending tasks, it is apparent there's a lot of pressure on this particular taskq.  My gut feeling is that this one patch won't by itself make a huge difference.  The big win will be the (not-yet-posted) patch to fan out the taskq on non-16k stack systems.  Watching the perf output with both patches on a non-16k stack system showed the time in raw_spin_lock_irq due to tq_lock vanish but, unfortunately, only to be replaced by extreme contention on another lock in the kernel (forgot which at the moment, but it's the one involved with disabling set-uid on files being written).  This was observed while trying to troubleshoot 3871 (intentionally not referenced) & friends.  There have been anecdotal reports that disabling dynamic taskqs helps some peoples' workloads and I have to wonder if this might be the cause (also not yet investigated).

**Comment by dweeezil** [comment]
My suspicions were correct.  Running an fio test:
```
[test]
	blocksize=8k
	scramble_buffers=1
	disk_util=0
	invalidate=0
	size=256m
	numjobs=32
	create_serialize=1
	direct=1
	filename=/dev/zvol/tank/v1
	offset=0
	offset_increment=10g
	group_reporting=1
	ioengine=libaio
	iodepth=10
	rw=write
	thread=1
	time_based=1
	runtime=1m
	fsync=0
```
which really exacerbates the issue, there were 917861 on tq_lock without this patch and it only dropped to 875989 with this patch.  However, when I fanned out the write issue taskq a bit (simply made it 12x5), the contentions on tq_lock dropped to 32763.  I did not record the actual fio results but they're not much different.  Fixing this point of contention simply moves it elsewhere, unfortunately.

I still think this patch is good on its own.  Among other things, it would allow us to use the same ASSERT in taskq_create that illumos has (disallowing cpu-proportioned taskqs to be dynamic).

The numbers above are for a stock 3.2 kernel which doesn't have 16K stacks.

I'm not sure what a good strategy would be yet to fan out the taskq in a mannter congruent with the concept of a cpu-proportioned taskq which is supposed to keep the whole system from melting down.  The situation is quite different between low core count systems and larger ones.

**Comment by behlendorf** [comment]
I'm OK with merging this small change as is even if it just moves the contention point since it's pretty clearly the right thing to do.  Just let me know how you want to proceed.

> I'm not sure what a good strategy would be yet to fan out the taskq in a manner congruent with the concept of a cpu-proportioned taskq which is supposed to keep the whole system from melting down. The situation is quite different between low core count systems and larger ones.

The first concern should probably be to verify that we can in fact fan this out safely.  Are we going to run in to problems by dispatching to different taskqs?  I don't foresee any issues here but it's something to keep in mind.

As for scaling this up perhaps it could be done by creating a taskq per cpu each with just a small number of threads.  We could create a new ZTI_* type for this type of workload or change the existing ZTI_MODE_BATCH implementation since this is the only existing consumer.

**Comment by dweeezil** [comment]
@behlendorf I'd like to see this merged as-is as a reasonable and correct first step.  A new ZTI type with a taskq per-cpu certainly makes sense but it's not "compatible" with the existing 75% scheme.   Maybe it's not be such a big deal now that we're actually running the write issue threads at a (every so slightly) lower priority.
